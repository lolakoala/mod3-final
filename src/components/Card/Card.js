import React, { Component } from 'react';
import { shape } from 'prop-types';
import './Card.css';
import fetchSwornMembers from './helper.js';

class Card extends Component {
  constructor() {
    super();
    this.state = {
      swornMembers: []
    };
  }
  handleClick = () => {
    const { swornMembers } = this.state;
    if (swornMembers.length >= 1) {
      this.setState({ swornMembers: [] });
      return;
    }
    const members = this.props.house.swornMembers.map(url => {
      return fetchSwornMembers(url)
        .then(res => res.json())
        .then(res => res.name);
    });
    Promise.all(members)
      .then(res => this.setState({ swornMembers: res }));
  }

  renderSwornMembers = () => {
    return this.state.swornMembers.map(member => <p key={member}>{member}</p>);
  }

  render() {
    const {
      name,
      founded,
      seats,
      titles,
      coatOfArms,
      ancestralWeapons,
      words,
      swornMembers
    } = this.props.house;

    const foundedInfo = founded.length ? founded : 'N/A';
    const titlesInfo = titles.length ?
      titles.map(title => <p
        className='title'
        key={title}>{`Title: ${title}`}
      </p>)
      : null;

    return (
      <div onClick={this.handleClick} className='Card'>
        <p className='name'>{name}</p>
        {words.length ? <p className='words'>{words}</p> : null}
        <p className='founded'>{`Founded: ${foundedInfo}`}</p>
        <p className='seats'>{`Seats: ${seats}`}</p>
        {titlesInfo}
        <p className='weapons'>{`Ancestral Weapons: ${ancestralWeapons}`}</p>
        <p className='coat'>{`Coat of Arms: ${coatOfArms}`}</p>
        {this.state.swornMembers.length === swornMembers.length ?
          <div className='members'>
            <p>Sworn Members:</p>
            {this.renderSwornMembers()}
          </div>
          : null}
      </div>
    );
  }
}

Card.propTypes = {
  house: shape.isRequired
};

export default Card;
