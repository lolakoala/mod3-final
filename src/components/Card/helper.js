const fetchSwornMembers = (url) => {
  fetch(url, {
    method: 'POST',
    body: JSON.stringify({ url: url })
  });
};

export default fetchSwornMembers;
