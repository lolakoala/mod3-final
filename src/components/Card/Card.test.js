import React from "react";
import Card from './Card.js';
import Enzyme, { shallow } from "enzyme";
import houses from '../../mockData/houses.js';
import Adapter from 'enzyme-adapter-react-16';

describe('card component', () => {
  Enzyme.configure({ adapter: new Adapter() });
  const wrapper = shallow(<Card house={houses[5]}/>);

  it('should be defined', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a name', () => {
    const name = wrapper.find('.name');

    expect(name.text()).toEqual(houses[5].name);
  });

  it('should have words', () => {
    const words = wrapper.find('.words');

    expect(words.text()).toEqual(houses[5].words);
  });

  it('should have founded info', () => {
    const founded = wrapper.find('.founded');

    expect(founded.text()).toEqual(`Founded: ${houses[5].founded}`);
  });

  it('should have seats', () => {
    const seats = wrapper.find('.seats');

    expect(seats.text()).toEqual(`Seats: ${houses[5].seats}`);
  });

  it('should have titles', () => {
    const titles = wrapper.find('.title');

    expect(titles.length).toEqual(houses[5].titles.length);
  });

  it('should have weapons', () => {
    const weapons = wrapper.find('.weapons');

    expect(weapons.text())
      .toEqual(`Ancestral Weapons: ${houses[5].ancestralWeapons}`);
  });

  it('should have a coat', () => {
    const coat = wrapper.find('.coat');

    expect(coat.text()).toEqual(`Coat of Arms: ${houses[5].coatOfArms}`);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
