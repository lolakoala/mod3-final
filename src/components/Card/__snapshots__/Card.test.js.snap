// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    house={
        Object {
            "ancestralWeapons": Array [
              "Lamentation",
            ],
            "cadetBranches": Array [
              "https://www.anapioficeandfire.com/api/houses/329",
            ],
            "coatOfArms": "Black iron studs on a bronze field, bordered with runes.",
            "currentLord": "https://www.anapioficeandfire.com/api/characters/1131",
            "diedOut": "",
            "founded": "Age of Heroes",
            "founder": "",
            "heir": "https://www.anapioficeandfire.com/api/characters/116",
            "name": "House Royce of Runestone",
            "overlord": "https://www.anapioficeandfire.com/api/houses/7",
            "region": "The Vale",
            "seats": Array [
              "Runestone",
            ],
            "swornMembers": Array [
              "https://www.anapioficeandfire.com/api/characters/116",
              "https://www.anapioficeandfire.com/api/characters/558",
              "https://www.anapioficeandfire.com/api/characters/608",
              "https://www.anapioficeandfire.com/api/characters/648",
              "https://www.anapioficeandfire.com/api/characters/818",
              "https://www.anapioficeandfire.com/api/characters/878",
              "https://www.anapioficeandfire.com/api/characters/892",
              "https://www.anapioficeandfire.com/api/characters/948",
              "https://www.anapioficeandfire.com/api/characters/1104",
              "https://www.anapioficeandfire.com/api/characters/1131",
              "https://www.anapioficeandfire.com/api/characters/1133",
              "https://www.anapioficeandfire.com/api/characters/1929",
            ],
            "titles": Array [
              "High King of the Vale, the Fingers and the Mountains of the Moon.",
              "Lord of Runestone",
            ],
            "url": "https://www.anapioficeandfire.com/api/houses/328",
            "words": "We Remember",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="name"
>
          House Royce of Runestone
</p>,
        <p
          className="words"
>
          We Remember
</p>,
        <p
          className="founded"
>
          Founded: Age of Heroes
</p>,
        <p
          className="seats"
>
          Seats: Runestone
</p>,
        Array [
          <p
            className="title"
>
            Title: High King of the Vale, the Fingers and the Mountains of the Moon.
</p>,
          <p
            className="title"
>
            Title: Lord of Runestone
</p>,
        ],
        <p
          className="weapons"
>
          Ancestral Weapons: Lamentation
</p>,
        <p
          className="coat"
>
          Coat of Arms: Black iron studs on a bronze field, bordered with runes.
</p>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "House Royce of Runestone",
          "className": "name",
        },
        "ref": null,
        "rendered": "House Royce of Runestone",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "We Remember",
          "className": "words",
        },
        "ref": null,
        "rendered": "We Remember",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Founded: Age of Heroes",
          "className": "founded",
        },
        "ref": null,
        "rendered": "Founded: Age of Heroes",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Seats: Runestone",
          "className": "seats",
        },
        "ref": null,
        "rendered": "Seats: Runestone",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": "High King of the Vale, the Fingers and the Mountains of the Moon.",
        "nodeType": "host",
        "props": Object {
          "children": "Title: High King of the Vale, the Fingers and the Mountains of the Moon.",
          "className": "title",
        },
        "ref": null,
        "rendered": "Title: High King of the Vale, the Fingers and the Mountains of the Moon.",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": "Lord of Runestone",
        "nodeType": "host",
        "props": Object {
          "children": "Title: Lord of Runestone",
          "className": "title",
        },
        "ref": null,
        "rendered": "Title: Lord of Runestone",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Ancestral Weapons: Lamentation",
          "className": "weapons",
        },
        "ref": null,
        "rendered": "Ancestral Weapons: Lamentation",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Coat of Arms: Black iron studs on a bronze field, bordered with runes.",
          "className": "coat",
        },
        "ref": null,
        "rendered": "Coat of Arms: Black iron studs on a bronze field, bordered with runes.",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="name"
>
            House Royce of Runestone
</p>,
          <p
            className="words"
>
            We Remember
</p>,
          <p
            className="founded"
>
            Founded: Age of Heroes
</p>,
          <p
            className="seats"
>
            Seats: Runestone
</p>,
          Array [
            <p
              className="title"
>
              Title: High King of the Vale, the Fingers and the Mountains of the Moon.
</p>,
            <p
              className="title"
>
              Title: Lord of Runestone
</p>,
          ],
          <p
            className="weapons"
>
            Ancestral Weapons: Lamentation
</p>,
          <p
            className="coat"
>
            Coat of Arms: Black iron studs on a bronze field, bordered with runes.
</p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "House Royce of Runestone",
            "className": "name",
          },
          "ref": null,
          "rendered": "House Royce of Runestone",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "We Remember",
            "className": "words",
          },
          "ref": null,
          "rendered": "We Remember",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Founded: Age of Heroes",
            "className": "founded",
          },
          "ref": null,
          "rendered": "Founded: Age of Heroes",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Seats: Runestone",
            "className": "seats",
          },
          "ref": null,
          "rendered": "Seats: Runestone",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": "High King of the Vale, the Fingers and the Mountains of the Moon.",
          "nodeType": "host",
          "props": Object {
            "children": "Title: High King of the Vale, the Fingers and the Mountains of the Moon.",
            "className": "title",
          },
          "ref": null,
          "rendered": "Title: High King of the Vale, the Fingers and the Mountains of the Moon.",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": "Lord of Runestone",
          "nodeType": "host",
          "props": Object {
            "children": "Title: Lord of Runestone",
            "className": "title",
          },
          "ref": null,
          "rendered": "Title: Lord of Runestone",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Ancestral Weapons: Lamentation",
            "className": "weapons",
          },
          "ref": null,
          "rendered": "Ancestral Weapons: Lamentation",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Coat of Arms: Black iron studs on a bronze field, bordered with runes.",
            "className": "coat",
          },
          "ref": null,
          "rendered": "Coat of Arms: Black iron studs on a bronze field, bordered with runes.",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
